File: articles/alonzo-church-lambda-calculus-and-formal-computation/typescript/church_numeral.ts
function churchApply<T>(n: number, f: (x: T) => T, x: T): T {
  for (let i = 0; i < n; i += 1) {
    x = f(x);
  }
  return x;
}

console.log(`church_3_successor_0=${churchApply(3, (x: number) => x + 1, 0)}`);

File: articles/alonzo-church-lambda-calculus-and-formal-computation/cpp/church_numeral.cpp
#include <iostream>
#include <functional>

int church_apply(int n, const std::function<int(int)>& f, int x) {
    for (int i = 0; i < n; ++i) {
        x = f(x);
    }
    return x;
}

int main() {
    std::cout << "church_3_successor_0=" << church_apply(3, [](int x){ return x + 1; }, 0) << "\n";
    return 0;
}
